src/dbnode/persist/fs/seek_manager.go

```diff
@@ -308,46 +538,79 @@ func (m *seekerManager) UpdateOpenLease(
 // and then notify the waiting goroutines that we've finished.
 func (m *seekerManager) getOrOpenSeekersWithLock(start xtime.UnixNano, byTime *seekersByTime) (seekersAndBloom, error) {
 	seekers, ok := byTime.seekers[start]
-	if ok && seekers.wg == nil {
+	if ok && seekers.active.wg == nil {
 		// Seekers are already open
-		return seekers, nil
+		return seekers.active, nil
 	}
 
-	if seekers.wg != nil {
+	if seekers.active.wg != nil {
 		// Seekers are being initialized / opened, wait for the that to complete
 		byTime.Unlock()
-		seekers.wg.Wait()
+		seekers.active.wg.Wait()
 		byTime.Lock()
 		// Need to do the lookup again recursively to see the new state
 		return m.getOrOpenSeekersWithLock(start, byTime)
 	}
 
-	// Seekers need to be opened
-	borrowableSeekers := make([]borrowableSeeker, 0, m.fetchConcurrency)
+	// Seekers need to be opened.
 	// We're going to release the lock temporarily, so we initialize a WaitGroup
 	// that other routines which would have otherwise attempted to also open this
 	// same seeker can use instead to wait for us to finish.
 	wg := &sync.WaitGroup{}
-	seekers.wg = wg
-	seekers.wg.Add(1)
+	seekers.active.wg = wg
+	seekers.active.wg.Add(1)
 	byTime.seekers[start] = seekers
 	byTime.Unlock()
+
+	onDoneOpeningNewSeeker := func() {
+		// Lock must be held when function returns.
+		byTime.Lock()
+		// Signal to other waiting goroutines that this goroutine is done attempting to open
+		// the seekers. This is done *after* acquiring the lock so that other goroutines that
+		// were waiting won't acquire the lock before this goroutine does.
+		wg.Done()
+	}
+	onErr := func() {
+		onDoneOpeningNewSeeker()
+		// Delete the seekersByTime struct so that the process can be restarted by the next
+		// goroutine (since this one errored out).
+		delete(byTime.seekers, start)
+	}
```

Comment:
Hm, this definitely is adding more hairy semantics inside an already complex function too.

Can we do something like:
```go
func (m *seekerManager) getOrOpenSeekerWithLock(start xtime.UnixNano, byTime *seekersByTime) (seekersAndBloom, error) {
	// all previous code

	newActiveSeekers, err := m.openSeekersWithActiveWaitGroup(start, byTime)

	// Re-establish lock and signal
	byTime.Lock()
	wg.Done()

	if err != nil {
		// Delete so process can be restarted...
		delete(byTime.seekers, start)
		return err
	}

	return seekersAndBloom, nil
}

func (m *seekerManager) openSeekersWithActiveWaitGroup(start xtime.UnixNano, byTime *seekersByTime) (seekersAndBloom, error) {
	// Open first one - Do this outside the context of the lock because opening
	// a seeker can be an expensive operation (validating index files).
	blm := m.blockRetrieverOpts.BlockLeaseManager()
	state, err := blm.OpenLatestLease(m, block.LeaseDescriptor{
		Namespace:  m.namespace,
		Shard:      byTime.shard,
		BlockStart: start.ToTime(),
	})
	if err != nil {
		return seekersAndBloom{}, fmt.Errorf("err opening latest lease: %v", err)
	}

	seeker, err := m.newOpenSeekerFn(byTime.shard, start.ToTime(), state.Volume)
	if err != nil {
		return seekersAndBloom{}, err
	}

	activeSeekers, err := m.seekersAndBloomFromSeeker(seeker, state.Volume)
	if err != nil {
		return seekersAndBloom{}, err
	}

	seekers.active = activeSeekers
	byTime.seekers[start] = seekers
	return activeSeekers, nil
}
```

Reply:
yeah looks good (or something similar at least once I test this out). Will ping once implemented

Reply:
OK made a few minor tweaks but this is implemented now (move the last two lines out of this function because they need to happen with the lock held)